src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +91,334 @@ data BlockRollbackResult
+            (fromIntegral $ begp ^. begpBlockEventCount)
+            (genBlockEvent begp)
+        finalEvent <- get <&> \case
+            GbesStartingState -> []
+            GbesExpectedFailure -> []
+            GbesBlockchain blockchain ->
+                -- Rollback the entire blockchain.
+                [BlkEvRollback $ BlockEventRollback
+                    { _berInput    = toNewestFirst blockchain
+                    , _berOutValid = BlockRollbackSuccess
+                    }]
+        return $ events ++ finalEvent
+
+replicateWhileM :: Monad m => Int -> m (Maybe a) -> m [a]
+replicateWhileM n m = go n
+  where
+    go k | k < 1 = return []
+         | otherwise =
+             m >>= \case
+                 Nothing -> return []
+                 Just a  -> (a:) <$> go (k-1)
+
+data RollbackFailureType
+    = RftRollbackExcess
+    | RftRollbackDrop
+    deriving (Enum, Bounded)
+
+instance Random RollbackFailureType where
+    random = randomR (minBound, maxBound)
+    randomR (lo,hi) = runRand $ uniform [lo..hi]
+
+data ApplyFailureType
+    = AftApplyBad
+    | AftApplyNonCont
+    deriving (Enum, Bounded)
+
+instance Random ApplyFailureType where
+    random = randomR (minBound, maxBound)
+    randomR (lo,hi) = runRand $ uniform [lo..hi]
+
+genBlockEvent ::
+       (Monad m, RandomGen g)
+    => BlockEventGenParams
+    -> StateT GenBlockEventState (RandT g m) (Maybe (BlockEvent' BlockIndex))
+genBlockEvent begp = do
+    gbes <- get
+    failure <- lift $ byChance (begp ^. begpFailureChance)
+    when failure $ put GbesExpectedFailure
+    rollback <- lift $ byChance (begp ^. begpRollbackChance)
+    case gbes of
+        GbesExpectedFailure -> return Nothing
+        GbesStartingState -> Just <$> case (failure, rollback) of
+            (True, True) -> do
+                -- Fail with rollback. In the starting state it's easy,
+                -- because any rollback will fail when we don't have any
+                -- blocks yet.
+                blockIndices <- genBlockIndices 0
+                return . BlkEvRollback $ BlockEventRollback
+                    { _berInput    = toNewestFirst blockIndices
+                    , _berOutValid = BlockRollbackFailure
+                    }
+            (True, False) -> do
+                -- Fail without rollback (with apply). In the starting state,
+                -- the only way to do this is to generate a sequence of blocks
+                -- invalid by itself.
+                blockIndices <- genBlockIndices 1
+                let
+                    blkZeroth     = BlockIndex 0 :| []
+                    blockIndices' =
+                        -- Those block indices are broken by construction
+                        -- because we append the 0-th block to the end.
+                        -- Sometimes we can violate the maximum bound on
+                        -- block count, but this is fine. See NOTE
+                        -- on reordering corner cases.
```

Comment:
I don't fully understand what these comments mean (I see a problem, then I see that the solution is proposed, but it has a drawback too and it's hard to understand what is done eventually), but do I understand correctly that we can generate more than `begpBlockCountMax` blocks? Even by 2.
1. `blockIndices` has at most `begpBlockCountMax` indices (see my comment to `genBlockIndices`).
2. We also add the 0-th block.

Update: I read this comment again and see this:
```
Sometimes we can violate the maximum bound on
                       -- block count, but this is fine.
```
Then I have another question: why is it fine?